### DISCUSSION TITLE
Where can I find the commands available from hooks

### DISCUSSION THREAD

**Discussion by lyndhurst** [qtile/qtile#4692] [Q&A] [answered]
Is there a reference listing the commands available when working on hooks ? Or can someone help me understand how to use the ones referenced in the documentation ?

I have been trying for a while to write a hook that would move focus to a group when a new window is created in it. I basically wanted the focus to follow windows moved on startup by my group `Matches`.

I finally succeeded with the following:

```python
@hook.subscribe.group_window_add
def group_window_add(group, window):
    if group.name != "scratchpad":
        group.cmd_toscreen(toggle=False)
```

In the documentation the only example in found in the hooks section is `client.to_group()` . This was enough to keep looking for a solution, but I could not figure out how to use something like `lazy.screen.toggle_group()` in a hook.

Finally after a lot searching I found `cmd_toscreen` in the discussions [here](https://github.com/qtile/qtile/discussions/2319#discussioncomment-536854).

 I would really appreciate if someone could point me towards an easier way to find the hook commands I am going to need in the future. Thanks.

**Comment by elParaguayo** [comment, marked as answer]
The "internal" commands are not in the documentation. However, to make it easier we've removed the use of the `cmd_` prefix so that the internal commands have the same names as the commands that you use for keybindings.

The main difference is getting access to the relevant python object. A good way to do this is to play around with `qtile shell` which lets you navigate qtile's command graph. You can then inspect objects with `eval(self)`.

For example:
```
 > cd group
group
group > eval(dir(self))
(True,
 "['__abstractmethods__', '__class__', '__delattr__', '__dict__', '__dir__', "
 "'__doc__', '__eq__', '__format__', '__ge__', '__getattr__', "
 "'__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', "
 "'__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', "
 "'__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', "
 "'__str__', '__subclasshook__', '__weakref__', '_abc_impl', "
 "'_command_object', '_commands', '_configure', '_get_command_signature', "
 "'_get_group', '_items', '_remove_from_focus_history', '_select', 'add', "
 "'command', 'commands', 'current_layout', 'current_window', 'custom_layout', "
 "'doc', 'eval', 'floating_layout', 'focus', 'focus_back', 'focus_by_index', "
 "'focus_by_name', 'focus_history', 'function', 'get_next_group', "
 "'get_previous_group', 'hide', 'info', 'info_by_name', 'items', 'label', "
 "'last_focused', 'layout', 'layout_all', 'layouts', 'mark_floating', 'name', "
 "'next_window', 'prev_window', 'qtile', 'remove', 'screen', "
 "'screen_affinity', 'select', 'set_label', 'set_screen', 'setlayout', "
 "'swap_window_order', 'switch_groups', 'tiled_windows', 'toscreen', "
 "'unminimize_all', 'use_layout', 'use_next_layout', 'use_previous_layout', "
 "'windows']")
```
Shows that the `group` object has a `screen` attribute.

```
group > eval(dir(self.screen))
(True,
 "['__abstractmethods__', '__annotations__', '__class__', '__delattr__', "
 "'__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', "
 "'__getattr__', '__getattribute__', '__getstate__', '__gt__', '__hash__', "
 "'__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', "
 "'__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', "
 "'__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', "
 "'_command_object', '_commands', '_configure', '_get_command_signature', "
 "'_items', '_select', '_toggle_group', 'bottom', 'command', 'commands', "
 "'dheight', 'doc', 'dwidth', 'dx', 'dy', 'eval', 'function', 'gaps', "
 "'get_rect', 'group', 'height', 'index', 'info', 'items', 'left', "
 "'next_group', 'paint', 'prev_group', 'previous_group', 'qtile', 'resize', "
 "'right', 'select', 'set_group', 'set_wallpaper', 'toggle_group', 'top', "
 "'wallpaper', 'wallpaper_mode', 'width', 'x', 'x11_drag_polling_rate', 'y']")
```
Reveals the `toggle_group` method.

So, you can do `group.screen.toggle_group()`. NB `screen` can be `None` so you should check that first.

**Reply by lyndhurst** [reply]
I am not sorry I asked this time ! I am making a note of this advice, it makes my testing workflow so much easier; I have been playing with the `qtile shell` this morning, this is exactly what I needed and more.

Now I can find commands, navigate the `graph` and hopefully at some point understand it better, but most importantly I can test commands instantly (on a VM to be safe) without having to edit the config file, `M-C-r` to restart qtile, trying my changes, and checking the logs when it is not working, I now have my error messages right away, this is a huge time saver.

I had seen mentions of the shell in the doc, but could not figure out how to use it right away, so I did not insist. Well, I should have.
Thank you @elParaguayo for the nth time now. I have been participating in those discussions for a very short time, but you consistently answer almost right away and in details. Makes me wonder if you managed to hack some kind of chatGPT assisted bot to help us qtile noobs because if those efforts are 100% human, I admire the expertise & commitment !

**Reply by elParaguayo** [reply]
You're most welcome. Keep asking as many questions as you need and we will help where we can.

**Comment by lyndhurst** [comment]
I have been trying to improve the code in the light of the answer I received.

The `cmd_` part of the command is indeed unnecessary, so the function now looks like that:

```python
@hook.subscribe.group_window_add
def group_window_add(group, window):
    if group.name != "scratchpad":
        group.toscreen(toggle=False)
```

On the other hand, I am not sure about @elParaguayo warning that `screen` can be `None`, and how to check it.

I have been trying to retrieve the group information dictionnary using `group_info = group.info()` and extracting the `"screen"` key to check if it exists. Since `screens` are 0 based integers, I used `isinstance(group_info["screen"], int)` as a conditional check.

The logic works, the problem is that the `group` object I can access within that function is the group that is receiving a new client and (in the case I am addressing here) does not have the focus, so the value for the `"screen"` is actually `None`, and the function works without errors.

So I am trying to figure out if I should check if there is a screen at all, but I do not manage to access the graph `root` or anything above the `group` passed as an argument to the function for that matter. I will come back with a better hook function when I figure it out.

**Reply by elParaguayo** [reply]
The comment about `screen` being `None` is only relevant if you're trying to access a method of the screen object (e.g. `group.screen.toggle_group`. If the group is not on a screen then you'd effectively be trying to call `None.toggle_group` which would result in an error.

The `group.toscreen` command has an optional `screen` argument. If you leave this blank, the group will be placed on the currently active screen. Alternatively, you can provide an integer which is the index of the particular screen you want to put the group on.

**Comment by lyndhurst** [comment]
Ok then, I am satisfied with the way things are working for now. I will keep all those infos in mind if I am trying to achieve some more elaborate multi screen setups in the future :+1: 

Thanks again.
